### PR TITLE
Improved tokenizing of context sensitive keywords

### PR DESCRIPTION
**Pull request by kukulich**
Other tokens should be added after these PR are merged:

- [ ] https://github.com/squizlabs/PHP_CodeSniffer/pull/3478
- [x] https://github.com/squizlabs/PHP_CodeSniffer/pull/3480
- [ ] https://github.com/squizlabs/PHP_CodeSniffer/pull/3483

### CODE REVIEW

**Comment by kukulich** [comment]
@gsherwood Please, this PR should be merged before https://github.com/squizlabs/PHP_CodeSniffer/pull/3478 and https://github.com/squizlabs/PHP_CodeSniffer/pull/3483.

It should really help in the future. I've already simplified the support for `T_READONLY`. I also cannot write some tests for `T_ENUM_CASE` without this PR.

File: src/Tokenizers/PHP.php [hunk 1]
```diff
@@ -1080,6 +1132,7 @@ protected function tokenize($string)
                 && PHP_VERSION_ID >= 50500
                 && $tokenIsArray === true
                 && $token[0] === T_YIELD
+                && isset($this->tstringContexts[$finalTokens[$lastNotEmptyToken]['code']]) === false
```
Comment (gsherwood) [inline src/Tokenizers/PHP.php:1135]: I'm not sure why this was added here. Removing it doesn't show any failing tests and I can't think of a code sample that would require this here.

File: src/Tokenizers/PHP.php [hunk 2]
```diff
@@ -589,6 +589,58 @@ protected function tokenize($string)
...
+                            $preserveKeyword = true;
+                        }
+
+                        break;
+                    }
+                }
+
+                if ($preserveKeyword === false) {
+                    if (PHP_CODESNIFFER_VERBOSITY > 1) {
+                        $type = $tokens[0];
```
Comment (gsherwood) [inline src/Tokenizers/PHP.php:629]: This should be `$token[0]` (not the `s`). This causes an error with verbosity on.

Comment (gsherwood) [reply src/Tokenizers/PHP.php:629]: Actually, this would be better as:
```
$type = Util\Tokens::tokenName($token[0]);
```
So that the token name shows up like in other verbose messages. It just shows an INT at the moment.

**Comment by gsherwood** [comment]
Left a couple of comments for minor stuff, but looks good to merge otherwise.

File: src/Tokenizers/PHP.php (hunk 2 above)
Comment (kukulich) [reply src/Tokenizers/PHP.php]: Fixed.

File: src/Tokenizers/PHP.php (hunk 1 above)
Comment (kukulich) [reply src/Tokenizers/PHP.php]: I'm not sure either. Removed.

File: src/Tokenizers/PHP.php [hunk 3]
```diff
@@ -589,6 +589,58 @@ protected function tokenize($string)
                 echo PHP_EOL;
             }
 
+            if ($tokenIsArray === true
```
Comment (jrfnl) [inline src/Tokenizers/PHP.php:592]: A comment above here to quickly see what this block is about would help.

**Review by jrfnl — COMMENTED** [review]
Nice one!, This would be a great step forward, but there's still quite a lot more to do in this regards (though not necessarily all in this PR). Also see: #3336 

Note: I locally made a start on a PR to handle this when I opened that issue, but got side-tracked before I could finish it as it is _hard_ to get that one right, especially when trying to account for all situations, including multi-use, group use etc.
Happy to share the code I have though.

---

Regarding this PR:

I wonder if there aren't more chunks of code in the tokenizer _after_ this new block, which could now be removed, or simplified ?

In particular, I'm looking at the "string-like token after a function keyword" block round line 1740 and the "special case for PHP 5.6 use function and use const" block round line 2168.

More extensive unit tests would also help catch edge cases.

File: src/Util/Tokens.php [hunk 4]
```diff
@@ -653,6 +653,83 @@ final class Tokens
         T_TRAIT_C  => T_TRAIT_C,
     ];
 
+    /**
+     * Tokens representing context sensitive keywords in PHP.
+     *
```
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Util/Tokens.php:658]: There will be PHPCS tokens in future, eg. `T_ENUM_CASE` but I'll check the tokens you mentioned.

[the remark this answers is not in the document]
Comment (kukulich) [reply src/Util/Tokens.php:658]: The test if failing when I remove the `T_PARENT` token.

```
1) PHP_CodeSniffer\Tests\Core\Tokenizer\ContextSensitiveKeywordsTest::testKeywords with data set #19 ('/* testParentIsKeyword */', 'T_PARENT')
Failed to find test target token for comment string: /* testParentIsKeyword */
Failed asserting that true is false.
```

File: src/Util/Tokens.php [hunk 5]
```diff
@@ -653,6 +653,83 @@ final class Tokens
...
+        T_PARENT       => T_PARENT,
+        T_PRINT        => T_PRINT,
+        T_PRIVATE      => T_PRIVATE,
+        T_PROTECTED    => T_PROTECTED,
+        T_PUBLIC       => T_PUBLIC,
+        T_READONLY     => T_READONLY,
+        T_REQUIRE      => T_REQUIRE,
+        T_REQUIRE_ONCE => T_REQUIRE_ONCE,
+        T_RETURN       => T_RETURN,
+        T_SELF         => T_SELF,
```
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Util/Tokens.php:720]:
```
1) PHP_CodeSniffer\Tests\Core\Tokenizer\ContextSensitiveKeywordsTest::testKeywords with data set #18 ('/* testSelfIsKeyword */', 'T_SELF')
Failed to find test target token for comment string: /* testSelfIsKeyword */
Failed asserting that true is false.
```

File: tests/Core/Tokenizer/ContextSensitiveKeywordsTest.php [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+
+use PHP_CodeSniffer\Tests\Core\AbstractMethodUnitTest;
+use PHP_CodeSniffer\Util\Tokens;
+
+class ContextSensitiveKeywordsTest extends AbstractMethodUnitTest
+{
+
+
+    /**
+     * TODO
```
[the remark this answers is not in the document]
Comment (kukulich) [reply tests/Core/Tokenizer/ContextSensitiveKeywordsTest.php]: Fixed, thanks.

File: tests/Core/Tokenizer/ContextSensitiveKeywordsTest.php [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+            ['/* testAnd */'],
+            ['/* testOr */'],
+            ['/* testXor */'],
+        ];
+
+    }//end dataStrings()
+
+
+    /**
+     * TODO
```
[the remark this answers is not in the document]
Comment (kukulich) [reply tests/Core/Tokenizer/ContextSensitiveKeywordsTest.php]: Fixed, thanks.

File: src/Tokenizers/PHP.php [hunk 8]
```diff
@@ -589,6 +589,58 @@ protected function tokenize($string)
...
+
+                // `new class` should be preserved
+                if ($token[0] === T_CLASS && $finalTokens[$lastNotEmptyToken]['code'] === T_NEW) {
+                    $preserveKeyword = true;
+                }
+
+                // `new class implements` should be preserved
+                if ($finalTokens[$lastNotEmptyToken]['code'] === T_CLASS) {
+                    $preserveKeyword = true;
+                }
```
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Tokenizers/PHP.php]: Improved the comment :)

File: src/Util/Tokens.php [hunk 9]
```diff
@@ -653,6 +653,83 @@ final class Tokens
...
+        T_INSTEADOF    => T_INSTEADOF,
+        T_INTERFACE    => T_INTERFACE,
+        T_LIST         => T_LIST,
+        T_LOGICAL_AND  => T_LOGICAL_AND,
+        T_LOGICAL_OR   => T_LOGICAL_OR,
+        T_LOGICAL_XOR  => T_LOGICAL_XOR,
+        T_MATCH        => T_MATCH,
+        T_NAMESPACE    => T_NAMESPACE,
+        T_NEW          => T_NEW,
+        T_PARENT       => T_PARENT,
```
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Util/Tokens.php:711]:
```
The test if failing when I remove the `T_PARENT` token.

```
1) PHP_CodeSniffer\Tests\Core\Tokenizer\ContextSensitiveKeywordsTest::testKeywords with data set #19 ('/* testParentIsKeyword */', 'T_PARENT')
Failed to find test target token for comment string: /* testParentIsKeyword */
Failed asserting that true is false.
```

File: src/Tokenizers/PHP.php (hunk 3 above)
Comment (kukulich) [reply src/Tokenizers/PHP.php:596]: Added.

**Comment by kukulich** [comment]
@jrfnl 

> In particular, I'm looking at the "string-like token after a function keyword" block round line 1740 and the "special case for PHP 5.6 use function and use const" block round line 2168.

"string-like token after a function keyword" is gone.

"special case for PHP 5.6 use function and use const" cannot be removed (without other work) but "This is a special case for the PHP 5.5 classname::class syntax" is gone.

File: src/Util/Tokens.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Util/Tokens.php:711]: @kukulich That's because it can't find token after the comment (as the custom token is no longer in the list), not because the tokenization is wrong.

File: src/Tokenizers/PHP.php [hunk 10]
```diff
@@ -588,6 +588,58 @@ protected function tokenize($string)
...
+
+                // `namespace\` should be preserved
```
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Tokenizers/PHP.php:612]: There's a test: https://github.com/squizlabs/PHP_CodeSniffer/pull/3484/commits/6f991b5a24d8ee36773a043c24f41156a0ed13ba#diff-884ebc6b1d508f7200a3ef0edc6dd662e97072d9fec37dd6a4e374a2d835f6d4R3-R4

File: src/Util/Tokens.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Util/Tokens.php:658]: In that case, `T_MATCH_DEFAULT` is missing too. And what about `T_CLOSURE` ?

All the same, for the check in the tokenizer where this list is used to look at the (non-empty) token _after_ the current one, it's not needed as the token _after_ hasn't been retokenized yet at that point.

File: src/Util/Tokens.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Util/Tokens.php:720]: See above

**Comment by jrfnl** [comment]
> but "This is a special case for the PHP 5.5 classname::class syntax" is gone.

Nice find! Yeah, of course, with the `T_PAAMAYIM_NEKUDOTAYIM` in the `$tstringContexts` that would be redundant.

File: src/Util/Tokens.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Util/Tokens.php:711]: @jrfnl Yes, you're right. However I'm not sure if the tokens should be removed here or not...

File: src/Tokenizers/PHP.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Tokenizers/PHP.php:612]: That's not what I meant (please unresolve this).

Think:
```php
namespace foreach;
namespace my\class\extensions;
```

[the remark this answers is not in the document]
Comment (kukulich) [reply src/Tokenizers/PHP.php:612]: Added test case.

File: src/Util/Tokens.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Util/Tokens.php:711]: I suppose that depends on where else this token list will be used, but for the current use case, those tokens aren't needed in the list.

[the remark this answers is not in the document]
Comment (kukulich) [reply src/Util/Tokens.php]: Removed for now.

File: src/Util/Tokens.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Util/Tokens.php:658]: It looks I was wrong. I don't need to add `T_ENUM_CASE` to the list. The same applies for `T_MATCH_DEFAULT `

File: src/Tokenizers/PHP.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Tokenizers/PHP.php:612]: Noted. That test case still only safeguards the first of the above examples though, the current tokenizer would still fail on the second.

File: src/Util/Tokens.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Util/Tokens.php:658]: Correct, as those won't be retokenized yet at the time the list is used, but as I mention in https://github.com/squizlabs/PHP_CodeSniffer/pull/3484#discussion_r773002627:

> I suppose that depends on where else this token list will be used, but for the current use case, those tokens aren't needed in the list.

File: src/Tokenizers/PHP.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Tokenizers/PHP.php:612]: Added one more test :)

File: src/Tokenizers/PHP.php [hunk 11]
```diff
@@ -1113,6 +1165,7 @@ protected function tokenize($string)
                 && $tokenIsArray === true
                 && $token[0] === T_STRING
                 && strtolower($token[1]) === 'yield'
+                && isset($this->tstringContexts[$finalTokens[$lastNotEmptyToken]['code']]) === false
```
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Tokenizers/PHP.php:1174]: `T_MATCH` can probably simplified but I'm not sure about this test case:

```
/* testLiveCoding */
// Intentional parse error. This has to be the last test in the file.
echo match
```

The token is currently tokenized as `T_STRING` but I'm not sure if it's right.

File: src/Tokenizers/PHP.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Tokenizers/PHP.php:614]: Thanks. Would be wonderful though if the other parts of the namespace name would also be checked. The test currently only checks the first part, i.e. `my`.

What I'm trying to point out, is that on PHP < 8.0, this snippet:
```php
<?php

namespace my\class\extensions;
namespace my\foreach\yield;
```

... would be tokenized as - take note of `class`, `yield` and `foreach`:
```
Ptr | Ln | Col  | Cond | ( #) | Token Type                 | [len]: Content
-------------------------------------------------------------------------
  0 | L1 | C  1 | CC 0 | ( 0) | T_OPEN_TAG                 | [5]: <?php

  1 | L2 | C  1 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

  2 | L3 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
  3 | L3 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
  4 | L3 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
  5 | L3 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  6 | L3 | C 14 | CC 0 | ( 0) | T_CLASS                    | [5]: class
  7 | L3 | C 19 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  8 | L3 | C 20 | CC 0 | ( 0) | T_STRING                   | [10]: extensions
  9 | L3 | C 30 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 10 | L3 | C 31 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

 11 | L4 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
 12 | L4 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
 13 | L4 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
 14 | L4 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 15 | L4 | C 14 | CC 0 | ( 0) | T_FOREACH                  | [7]: foreach
 16 | L4 | C 21 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 17 | L4 | C 22 | CC 0 | ( 0) | T_YIELD                    | [5]: yield
 18 | L4 | C 27 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 19 | L4 | C 28 | CC 0 | ( 0) | T_WHITESPACE               | [0]:
```

... while with your change, it will be tokenized (better), like so:
```
Ptr | Ln | Col  | Cond | ( #) | Token Type                 | [len]: Content
-------------------------------------------------------------------------
  0 | L1 | C  1 | CC 0 | ( 0) | T_OPEN_TAG                 | [5]: <?php

  1 | L2 | C  1 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

  2 | L3 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
  3 | L3 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
  4 | L3 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
  5 | L3 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  6 | L3 | C 14 | CC 0 | ( 0) | T_STRING                   | [5]: class
  7 | L3 | C 19 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  8 | L3 | C 20 | CC 0 | ( 0) | T_STRING                   | [10]: extensions
  9 | L3 | C 30 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 10 | L3 | C 31 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

 11 | L4 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
 12 | L4 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
 13 | L4 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
 14 | L4 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 15 | L4 | C 14 | CC 0 | ( 0) | T_STRING                   | [7]: foreach
 16 | L4 | C 21 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 17 | L4 | C 22 | CC 0 | ( 0) | T_STRING                   | [5]: yield
 18 | L4 | C 27 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 19 | L4 | C 28 | CC 0 | ( 0) | T_WHITESPACE               | [0]:
```

There is no difference on PHP 8.0+ as the "retokenization of namespaced names" already takes care of it in that case.

[the remark this answers is not in the document]
Comment (kukulich) [reply src/Tokenizers/PHP.php:614]: Ok, now I finally get it :) Added.

**Comment by gsherwood** [comment]
Just an update that I'm trying to merge this at the moment. I'm needing to resolve a lot of conflicts in the 4.0 branch first, and adjust the namespace changes over there as well.

**Comment by gsherwood** [comment]
Finally got this merged and tested after Github had issues today, but all working.

Appreciate all the work getting this done. Keen to have anyone else run their eyes over things as well to make sure I haven't missed something obvious.

**Comment by jrfnl** [comment]
> Keen to have anyone else run their eyes over things as well to make sure I haven't missed something obvious.

If you like, I'll try to have a look at the 4.x branch later in the week ? (on that note: there are still a few PRs open for that branch)
